iot-hub/Samples/device/DeviceReconnectionSample/DeviceReconnectionSample.cs

```diff
@@ -231,37 +258,102 @@ private async Task SendMessagesAsync(CancellationToken cancellationToken)
+                            await Task.Delay(s_sleepDuration);
+                            continue;
+                        }
+
+                        await CloudToDeviceMessageHandler(receivedMessage, null);
+                    }
+                    catch (IotHubException ex) when (ex.IsTransient)
+                    {
+                        _logger.LogError($"A transient IotHubException was caught, will retry receiving the message: {ex}");
+                    }
+                    catch (Exception ex) when (ExceptionHelper.IsNetworkExceptionChain(ex))
+                    {
+                        _logger.LogError($"A network related exception was caught, will retry receiving the message: {ex}");
+                    }
+                }
+
+                await Task.Delay(s_sleepDuration);
+            }
+        }
 
-                _logger.LogInformation($"Device waiting for C2D messages from the hub for {s_sleepDuration}...");
-                _logger.LogInformation("Use the IoT Hub Azure Portal or Azure IoT Explorer to send a message to this device.");
+        private Task CloudToDeviceMessageHandler(Message receivedMessage, object context)
+        {
+            string messageData = Encoding.ASCII.GetString(receivedMessage.GetBytes());
+            var formattedMessage = new StringBuilder($"Received message: [{messageData}], with lockToken: [{receivedMessage.LockToken}].\n");
 
-                try
-                {
-                    await ReceiveMessageAndCompleteAsync();
-                }
-                catch (DeviceMessageLockLostException ex)
-                {
-                    _logger.LogWarning($"Attempted to complete a received message whose lock token has expired; ignoring: {ex}");
-                }
-                catch (IotHubException ex) when (ex.IsTransient)
-                {
-                    // Inspect the exception to figure out if operation should be retried, or if user-input is required.
-                    _logger.LogError($"An IotHubException was caught, but will try to recover and retry explicitly: {ex}");
-                }
-                catch (Exception ex) when (ExceptionHelper.IsNetworkExceptionChain(ex))
-                {
-                    _logger.LogError($"A network related exception was caught, but will try to recover and retry explicitly: {ex}");
-                }
+            foreach (var prop in receivedMessage.Properties)
+            {
+                formattedMessage.AppendLine($"\tProperty: key={prop.Key}, value={prop.Value}");
             }
+            _logger.LogInformation(formattedMessage.ToString());
+
+            _receivedMessagesQueue.Enqueue(receivedMessage);
+            return Task.CompletedTask;
+        }
+
+        private async Task CompleteReceivedMessagesAsync(CancellationToken cancellationToken)
+        {
+            while (!cancellationToken.IsCancellationRequested)
+            {
+                if (IsDeviceConnected)

```

Comment:
To prevent such massive indenting, how about handling error conditions different?

Like:

```csharp
if (!IsDeviceConnected)
{
    continue;
}
```

That way the happy path code doesn't need to be indented quite so many times?

You'd probably need to move the Task.Delay to the top of the loop though, then.